File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-0.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        memo = {}

        def opt(i, j):
            if (i, j) in memo:
                return memo[(i, j)]

            res = 1
            if i + 1 < len(matrix) and matrix[i + 1][j] > matrix[i][j]:
                res = max(res, 1 + opt(i + 1, j))

            if j + 1 < len(matrix[0]) and matrix[i][j + 1] > matrix[i][j]:
                res = max(res, 1 + opt(i, j + 1))
            
            if i - 1 >= 0 and matrix[i - 1][j] > matrix[i][j]:
                res = max(res, 1 + opt(i - 1, j))
            
            if j - 1 >= 0 and matrix[i][j - 1] > matrix[i][j]:
                res = max(res, 1 + opt(i, j - 1))
            
            memo[(i, j)] = res

            return res
        
        result = 0
        for i in range(len(matrix)):
            for j in range(len(matrix[i])):
                result = max(result, opt(i, j))

        return result




        

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        map<char, int> count;
        int res = 0;

        int l = 0, maxl = 0;

        for(int r = 0; r < s.size(); r++){
            count[s[r]]++;
            maxl = max(maxl, count[s[r]]);

            while((r - l + 1) - maxl > k){
                count[s[l]]--;
                l++;
            }

            res = max(res, r - l + 1);
        }

        return res;
    }
};
